# NAIVE BAYES


## 2.0 modelo de regressión bates ingenuo


In [14]:
import numpy as np
import pandas as pd
from sklearn.naive_bayes import GaussianNB
from sklearn.model_selection import train_test_split
from sklearn.metrics import accuracy_score, confusion_matrix, classification_report
from sklearn.impute import SimpleImputer
#se usa el gausian NB dado que la data no es discreta , sino continua
data = pd.read_csv('train.csv')
df = pd.read_csv('train.csv')
numeric_features = data.select_dtypes(include=['int64', 'float64']).columns.drop(['Id','SalePrice'])
#mimsa data de modelos lineales, y de arbol
X = df[numeric_features]
y = df['SalePrice']
imputer = SimpleImputer(strategy='mean')
X_imputed = imputer.fit_transform(X)


X_train, X_test, y_train, y_test = train_test_split(X_imputed, y, test_size=0.2, random_state=42)

nb_predictor = GaussianNB()
nb_predictor.fit(X_train, y_train)
y_pred = nb_predictor.predict(X_test)

accuracy_report = accuracy_score(y_test,y_pred)


print(f"Accuracy of model: {accuracy_report*100}%")


Accuracy of model: 0.684931506849315%


## 3. Analice los resultados del modelo de regresión. ¿Qué tan bien le fue prediciendo?

El resultado es pésimo , no es del 1%, pues el modelo requiere más preparación de los datos y naive bayes es para clasificacion no para predecir una variable de respuesta directamente

## 4 Compare los resultados con el modelo de regresión lineal y el árbol de regresión que hizo en las hojas pasadas. ¿Cuál funcionó mejor?

el que funcionó mejor es la regresión lineal con una precisión del 80% (R^2) , pues es el más optimo para predicciones de valores de una variable respuesta, mientras que naive bayes y arbol es para clasificacion, sin embargo el arbol obtuvo un accuracy de 72%

# 5 Haga un modelo de clasificación, use la variable categórica que hizo con el precio de las  casas (barata, media y cara) como variable respuesta.

In [4]:
import pandas as pd
from sklearn.model_selection import train_test_split
from sklearn.naive_bayes import GaussianNB
from sklearn.metrics import accuracy_score, confusion_matrix, classification_report
from sklearn.preprocessing import OneHotEncoder, StandardScaler
from sklearn.impute import SimpleImputer


data = pd.read_csv('train.csv')

# Creación de la variable categórica 'PriceCategory'
p25 = data['SalePrice'].quantile(0.25)
p75 = data['SalePrice'].quantile(0.75)
data['PriceCategory'] = pd.cut(data['SalePrice'], bins=[0, p25, p75, float('inf')], labels=['Económicas', 'Intermedias', 'Caras'], right=False)

# Preprocesamiento de variables categóricas
categorical_features = data.select_dtypes(include=['object']).columns
one_hot_encoder = OneHotEncoder(sparse=False, drop='first')
encoded_categorical_data = pd.DataFrame(one_hot_encoder.fit_transform(data[categorical_features]))
encoded_categorical_data.columns = one_hot_encoder.get_feature_names_out(categorical_features)

# Concatenación de características numéricas y categóricas codificadas
numeric_features = data.select_dtypes(include=['int64', 'float64']).columns.drop('SalePrice')
X = pd.concat([data[numeric_features], encoded_categorical_data], axis=1)
y = data['PriceCategory']

# Imputación de valores faltantes
numeric_imputer = SimpleImputer(strategy='mean')
categorical_imputer = SimpleImputer(strategy='most_frequent')

X[numeric_features] = numeric_imputer.fit_transform(X[numeric_features])
X[encoded_categorical_data.columns] = categorical_imputer.fit_transform(X[encoded_categorical_data.columns])

# División de los datos en conjuntos de entrenamiento y prueba
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)


model = GaussianNB()
model.fit(X_train, y_train)


y_pred = model.predict(X_test)

accuracy = accuracy_score(y_test, y_pred)
conf_matrix = confusion_matrix(y_test, y_pred)
class_report = classification_report(y_test, y_pred)

print(f'Accuracy: {accuracy}')
print('Confusion Matrix:')
print(conf_matrix)
print('Classification Report:')
print(class_report)


Accuracy: 0.6917808219178082
Confusion Matrix:
[[64  0  6]
 [ 0 76  5]
 [35 44 62]]
Classification Report:
              precision    recall  f1-score   support

       Caras       0.65      0.91      0.76        70
  Económicas       0.63      0.94      0.76        81
 Intermedias       0.85      0.44      0.58       141

    accuracy                           0.69       292
   macro avg       0.71      0.76      0.70       292
weighted avg       0.74      0.69      0.67       292



c:\Users\marce\AppData\Local\Programs\Python\Python312\Lib\site-packages\sklearn\preprocessing\_encoders.py:975: FutureWarning: `sparse` was renamed to `sparse_output` in version 1.2 and will be removed in 1.4. `sparse_output` is ignored unless you leave `sparse` to its default value.
  warnings.warn(


# 6 Utilice los modelos con el conjunto de prueba y determine la eficiencia del algoritmo para  predecir y clasificar. 

# Comparación de Modelos de Clasificación

## Modelo de Clasificación (Categorías: Baratas, Intermedias, Caras)

- **Precisión (Accuracy):** 69.18%
- **Matriz de confusión:**
[[64 0 6]
[ 0 76 5]
[35 44 62]]



### Informe de Clasificación
- Clase "Caras": precisión de 0.65, recall de 0.91.
- Clase "Económicas": precisión de 0.63, recall de 0.94.
- Clase "Intermedias": precisión de 0.85, recall de 0.44.

## Modelo de Naive Bayes Adaptado para Regresión (Categorías: Baratas, Intermedias, Caras)

- **Precisión (Accuracy):** 56.16%
- **Matriz de confusión:**
[[51 43 16]
[ 5 25 58]
[ 1 5 88]]



### Informe de Clasificación
- Clase "Baratas": precisión de 0.89, recall de 0.46.
- Clase "Intermedias": precisión de 0.34, recall de 0.28.
- Clase "Caras": precisión de 0.54, recall de 0.94.

## Comparación y Análisis

- El **modelo de clasificación** tiene una **precisión general más alta (69.18%)** en comparación con el **modelo adaptado para regresión (56.16%)**.
- En el modelo de clasificación, las clases "Caras" y "Económicas" tienen altos valores de recall, lo que indica que el modelo es bueno para identificar estas categorías. Sin embargo, la clase "Intermedias" tiene un recall bajo, lo que sugiere que el modelo tiene dificultades para identificar correctamente esta categoría.
- En el modelo adaptado para regresión, la clase "Caras" tiene un recall muy alto (0.94), pero la clase "Intermedias" tiene valores bajos tanto en precisión como en recall, lo que indica una dificultad significativa para clasificar correctamente esta categoría.
- Ambos modelos tienen un rendimiento variable en diferentes categorías, pero el **modelo de clasificación parece ser más equilibrado** en términos de precisión y recall para las tres categorías.

El **modelo de clasificación** es **más eficiente para predecir y clasificar** , especialmente para las categorías "Caras" y "Económicas". El modelo adaptado para regresión tiene un rendimiento más bajo,haciendo enfásis  en la categoría "Intermedias".


# 7 Haga un análisis de la eficiencia del modelo de clasificación usando una matriz de confusión. Tenga en cuenta la efectividad, donde el algoritmo se equivocó más, donde se equivocó menos y la importancia que tienen los errores. 

[[64  0  6]
 [ 0 76  5]
 [35 44 62]]


# Análisis Detallado de la Matriz de Confusión

## Clase "Caras"
- **Verdaderos Positivos (TP) = 64:** El modelo clasificó correctamente 64 casas como "Caras".
- **Falsos Negativos (FN) = 6:** El modelo no identificó 6 casas que eran "Caras".
- **Falsos Positivos (FP) = 0:** El modelo no clasificó incorrectamente ninguna casa como "Cara".
- **Análisis:** La clase "Caras" tiene una alta precisión y un buen recall, lo que indica que el modelo es muy confiable al identificar casas "Caras" y rara vez las confunde con otras categorías.

## Clase "Económicas"
- **Verdaderos Positivos (TP) = 76:** El modelo clasificó correctamente 76 casas como "Económicas".
- **Falsos Negativos (FN) = 5:** El modelo no identificó 5 casas que eran "Económicas".
- **Falsos Positivos (FP) = 0:** El modelo no clasificó incorrectamente ninguna casa como "Económica".
- **Análisis:** La clase "Económicas" tiene una alta precisión y un buen recall, lo que indica que el modelo es muy efectivo al clasificar casas en esta categoría.

## Clase "Intermedias"
- **Verdaderos Positivos (TP) = 62:** El modelo clasificó correctamente 62 casas como "Intermedias".
- **Falsos Negativos (FN) = 79:** El modelo no identificó 79 casas que eran "Intermedias".
- **Falsos Positivos (FP) = 79:** El modelo clasificó incorrectamente 79 casas como "Intermedias".
- **Análisis:** La clase "Intermedias" presenta el mayor desafío para el modelo, con una precisión más baja y un recall moderado. Esto indica una considerable confusión entre esta categoría y las otras.

# Importancia de los Errores

## Impacto de los Falsos Positivos (FP)
- Los FP en la clase "Intermedias" pueden llevar a una sobreestimación de la calidad o el valor de las casas. Esto podría resultar en decisiones de inversión equivocadas.

## Impacto de los Falsos Negativos (FN)
- Los FN en la clase "Intermedias" pueden tener el efecto contrario, subestimando el valor de las casas que realmente pertenecen a esta categoría. Esto podría resultar en oportunidades perdidas o en la subvaloración de propiedades.






 # 8 Analice el modelo. ¿Cree que pueda estar sobre ajustado? 

In [5]:


from sklearn.metrics import accuracy_score

model = GaussianNB()
model.fit(X_train, y_train)

y_train_pred = model.predict(X_train)


train_accuracy = accuracy_score(y_train, y_train_pred)

print(f'Accuracy en el conjunto de entrenamiento: {train_accuracy}')
print(f'Accuracy en el conjunto de prueba: {accuracy}')


Accuracy en el conjunto de entrenamiento: 0.7243150684931506
Accuracy en el conjunto de prueba: 0.6917808219178082


Precisión en el conjunto de entrenamiento: 72.43%
Precisión en el conjunto de prueba: 69.18%
La diferencia entre la precisión en el conjunto de entrenamiento y la precisión en el conjunto de prueba es relativamente pequeña ,aproximadamente 3.25%. Esto indica que el modelo está generalizando bien a los datos no vistos y no muestra signos evidentes de sobreajuste

In [9]:
from sklearn.model_selection import cross_val_score


cv_scores = cross_val_score(model, X, y, cv=5, scoring='accuracy')

print(f'Puntajes de validación cruzada: {cv_scores}')
print(f'Promedio de puntajes de validación cruzada: {cv_scores.mean()}')


print(f'Accuracy del modelo anterior en el conjunto de prueba: {accuracy}')


Puntajes de validación cruzada: [0.6369863  0.69178082 0.73630137 0.70205479 0.71232877]
Promedio de puntajes de validación cruzada: 0.695890410958904
Accuracy del modelo anterior en el conjunto de prueba: 0.6917808219178082


El promedio de los puntajes de validación cruzada es ligeramente superior a la precisión del modelo anterior en el conjunto de prueba. La diferencia es pequeña, sin embargo, el modelo evaluado con validación cruzada funcionó ligeramente mejor

#  Tanto para los modelos de regresión como de clasificación,pruebe con varios valores de los hiperparámetros, use el mejor modelo del tunneo, ¿Mejoraron los modelos? Explique 

## Modelo de regresión ingenuo

In [2]:
import numpy as np
import pandas as pd
from sklearn.naive_bayes import GaussianNB
from sklearn.model_selection import train_test_split, GridSearchCV
from sklearn.metrics import accuracy_score
from sklearn.impute import SimpleImputer

df = pd.read_csv('train.csv')

numeric_features = df.select_dtypes(include=['int64', 'float64']).columns.drop(['Id', 'SalePrice'])

X = df[numeric_features]
y = df['SalePrice']

imputer = SimpleImputer(strategy='mean')
X_imputed = imputer.fit_transform(X)

X_train, X_test, y_train, y_test = train_test_split(X_imputed, y, test_size=0.2, random_state=42)

param_grid = {'var_smoothing': np.logspace(0,-9, num=100)}

grid_search = GridSearchCV(estimator=GaussianNB(), param_grid=param_grid, cv=5, scoring='accuracy')
grid_search.fit(X_train, y_train)

best_params = grid_search.best_params_

best_var_smoothing = best_params['var_smoothing']
best_nb_model = GaussianNB(var_smoothing=best_var_smoothing)
best_nb_model.fit(X_train, y_train)

# Realizar predicciones en los datos de prueba
y_pred_test = best_nb_model.predict(X_test)

# Calcular la precisión del modelo
accuracy = accuracy_score(y_test, y_pred_test)
print(f"Precisión del modelo con los mejores hiperparámetros: {accuracy*100}%")


c:\Python311\Lib\site-packages\sklearn\model_selection\_split.py:737: UserWarning: The least populated class in y has only 1 members, which is less than n_splits=5.
  warnings.warn(


Precisión del modelo con los mejores hiperparámetros: 1.0273972602739725%


In [11]:
import pandas as pd
from sklearn.model_selection import train_test_split
from sklearn.naive_bayes import GaussianNB
from sklearn.metrics import accuracy_score, confusion_matrix, classification_report
from sklearn.preprocessing import OneHotEncoder, StandardScaler
from sklearn.impute import SimpleImputer
from sklearn.pipeline import Pipeline
from sklearn.compose import ColumnTransformer 
# Leer el conjunto de datos
data = pd.read_csv('train.csv')

# Crear la variable categórica 'PriceCategory'
p25 = data['SalePrice'].quantile(0.25)
p75 = data['SalePrice'].quantile(0.75)
data['PriceCategory'] = pd.cut(data['SalePrice'], bins=[0, p25, p75, float('inf')], labels=['Económicas', 'Intermedias', 'Caras'], right=False)



# Definir características categóricas excluyendo 'PriceCategory'
categorical_features = data.select_dtypes(include=['object']).columns

# Definir características numéricas
numeric_features = data.select_dtypes(include=['int64', 'float64']).columns.drop(['Id', 'SalePrice'])

# Preprocesamiento de características numéricas
numeric_transformer = Pipeline(steps=[
    ('imputer', SimpleImputer(strategy='mean')),
    ('scaler', StandardScaler())
])

# Preprocesamiento de características categóricas
categorical_transformer = Pipeline(steps=[
    ('imputer', SimpleImputer(strategy='most_frequent')),
    ('onehot', OneHotEncoder(drop='first'))
])

# Combinar preprocesadores
preprocessor = ColumnTransformer(
    transformers=[
        ('num', numeric_transformer, numeric_features),
        ('cat', categorical_transformer, categorical_features)
    ])

X = df[numeric_features]
y = df['SalePrice']

imputer = SimpleImputer(strategy='mean')
X_imputed = imputer.fit_transform(X)

X_train, X_test, y_train, y_test = train_test_split(X_imputed, y, test_size=0.2, random_state=42)

# Modelo Naive Bayes Gaussiano
model = GaussianNB()

# Entrenamiento del modelo
model.fit(X_train, y_train)

# Predicción en el conjunto de prueba
y_pred = model.predict(X_test)

# Evaluación del modelo
accuracy = accuracy_score(y_test, y_pred)
conf_matrix = confusion_matrix(y_test, y_pred)
class_report = classification_report(y_test, y_pred)

print(f'Accuracy: {accuracy}')
print('Confusion Matrix:')
print(conf_matrix)
print('Classification Report:')
print(class_report)


Accuracy: 0.00684931506849315
Confusion Matrix:
[[0 0 0 ... 0 0 0]
 [0 0 0 ... 0 0 0]
 [0 0 0 ... 0 0 0]
 ...
 [0 0 0 ... 0 0 0]
 [0 0 0 ... 0 0 0]
 [0 0 0 ... 0 0 0]]
Classification Report:
              precision    recall  f1-score   support

       35311       0.00      0.00      0.00         1
       40000       0.00      0.00      0.00         1
       55000       0.00      0.00      0.00         0
       55993       0.00      0.00      0.00         1
       60000       0.00      0.00      0.00         1
       64500       0.00      0.00      0.00         1
       66500       0.00      0.00      0.00         1
       67000       0.00      0.00      0.00         2
       68400       0.00      0.00      0.00         1
       68500       0.00      0.00      0.00         1
       75000       0.00      0.00      0.00         1
       75500       0.00      0.00      0.00         1
       79000       0.00      0.00      0.00         0
       79500       0.00      0.00      0.00         

c:\Python311\Lib\site-packages\sklearn\metrics\_classification.py:1497: UndefinedMetricWarning: Precision is ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, f"{metric.capitalize()} is", len(result))
c:\Python311\Lib\site-packages\sklearn\metrics\_classification.py:1497: UndefinedMetricWarning: Recall is ill-defined and being set to 0.0 in labels with no true samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, f"{metric.capitalize()} is", len(result))
c:\Python311\Lib\site-packages\sklearn\metrics\_classification.py:1497: UndefinedMetricWarning: Precision is ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, f"{metric.capitalize()} is", len(result))
c:\Python311\Lib\site-packages\sklearn\metrics\_classification.py:1497: UndefinedMet

El modelo de clasificacion no mejora, sino empeora perdiendo toda su accuracy. Y el de regresión aumenta a 1.02%, pero igualmente es insignificante sus predicciones, pues no es viable.

# Compare la eficiencia del algoritmo con el resultado obtenido con el árbol de decisión (el de clasificación) y el modelo de random forest que hizo en la hoja pasada. ¿Cuál es mejor para predecir? ¿Cuál se demoró más en procesar? 

In [13]:
import time

start_time = time.time()
model.fit(X_train, y_train)
end_time = time.time()
print(f"Tiempo de entrenamiento: {end_time - start_time} segundos")


Tiempo de entrenamiento: 0.011999130249023438 segundos


Rendimiento: El modelo de Random Forest para regresión tiene un R² de 0.8899, lo que indica un buen ajuste al conjunto de datos. En comparación, el modelo Naive Bayes para clasificación tiene una precisión del 69.18%